src/core/ext/filters/http/server/http_server_filter.cc

```diff
@@ -306,19 +298,11 @@ static grpc_error_handle hs_filter_incoming_metadata(grpc_call_element* elem,
     }
   }
 
-  if (b->legacy_index()->named.host != nullptr &&
-      b->legacy_index()->named.authority == nullptr) {
-    grpc_linked_mdelem* el = b->legacy_index()->named.host;
-    grpc_mdelem md = GRPC_MDELEM_REF(el->md);
-    b->Remove(el);
-    hs_add_error(
-        error_name, &error,
-        grpc_metadata_batch_add_head(
-            b, el,
-            grpc_mdelem_from_slices(GRPC_MDSTR_AUTHORITY,
-                                    grpc_slice_ref_internal(GRPC_MDVALUE(md))),
-            GRPC_BATCH_AUTHORITY));
-    GRPC_MDELEM_UNREF(md);
+  if (b->legacy_index()->named.authority == nullptr) {
+    absl::optional<grpc_core::Slice> host = b->Take(grpc_core::HostMetadata());
+    if (host.has_value()) {
+      b->Append(":authority", std::move(*host));
```

Comment:
Should this be added to head instead of tail here (with Append) ? From my understanding, it appears the old code adds to head

Reply:
You're right, and `:` prefixed metadata should absolutely go to head per the http2 spec.

That said - this is the server filter, and so this metadata never goes out on the wire (http2 considerations are moot) and we loose the ability to reconstruct this specific order by the time this metadata *could* hit a wire.

Additionally, we don't have a `Prepend` function just yet that would work.

Now we could `Append` and stick some comment here, and that might be good - but within the next two changes in this series this code gets removed and we do `b->Set(AuthorityMetadata(), std::move(*host));`.

Similarly we could add `Prepend` and modify this code to use that and be completely the same behavior as before - but I'm not convinced we need `Prepend` once these changes are done and we do the aforementioned `b->Set(AuthorityMetadata(), std::move(*host));`